[ConstraintSolver] NFC: refactor candidate solver to use constraint system arena

xedin opened the pull request:
While shrinking we have to allocate containers for the reduced domains
for some of the candidates, it's currently done using permanent arena
of the `ASTContext` allocator. This patch changes candidate solver to
use arena associated with the parent constraint system, which significantly
limits lifetime of domain containers.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
/cc @jrose-apple mostly changes in the `ASTContext`

xedin commented:
@swift-ci please smoke test

jrose-apple commented on lib/Sema/CSSolver.cpp:
```diff
@@ -922,8 +922,9 @@ void ConstraintSystem::Candidate::applySolutions(
-      = TC.Context.AllocateUninitialized<ValueDecl *>(choices.size());
+    MutableArrayRef<ValueDecl *> decls =
+        TC.Context.AllocateUninitialized<ValueDecl *>(choices.size(),
+                                                      Allocator);
```
I think I would just skip the whole change for ASTContext and use the BumpPtrAllocator directly. You should be able to do this:

```c++
MutableArrayRef<ValueDecl *> decls(Allocator.Allocate<ValueDecl *>(choices.size()),
                                   choices.size());
```

xedin replied:
Done! Thanks for the suggestion!

xedin committed:
[ConstraintSolver] NFC: refactor candidate solver to use constraint system arena

While shrinking we have to allocate containers for the reduced domains
for some of the candidates, it's currently done using permanent arena
of the `ASTContext` allocator. This patch changes candidate solver to
use arena associated with the parent constraint system, which significantly
limits lifetime of domain containers.

xedin commented:
@swift-ci please clean smoke test

xedin commented:
@swift-ci please clean smoke test

xedin commented:
@swift-ci please smoke test

jrose-apple commented on lib/Sema/CSSolver.cpp:
```diff
@@ -922,8 +922,10 @@ void ConstraintSystem::Candidate::applySolutions(
+                                choices.size());
+
     std::uninitialized_copy(choices.begin(), choices.end(), decls.begin());
     OSR->setDecls(decls);
```
What keeps this from outliving the constraint system?

xedin replied:
If expression is correct there is not going to be any `OSR` expressions in it after `ExprRewriter` because all of the ambiguities have been resolved. But I think the right fix is to allocate to constraint system arena all of the intermediate arrays that `shrink` creates but at the end of the `shrink` phrase re-allocate all of the _final_ arrays to perm arena to make sure that diagnostics are ok.

jrose-apple replied:
Right, that first part isn't good enough for code completion and indexing, where the expression may not be complete or correct.

xedin commented:
@swift-ci please smoke test

xedin commented:
@rudkx all of the intermediate arrays used for decls are going to be allocated temporarily while main constraint system is alive but the final versions are going to be still permanent because we might need to keep OSR longer for diagnostics.

rudkx reviewed (approved):
LGTM!
